packages/storage/amplify_storage_s3_dart/lib/src/platform_impl/download_file/download_file_io.dart

```diff
@@ -0,0 +1,128 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:io';
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:amplify_storage_s3_dart/amplify_storage_s3_dart.dart';
+import 'package:amplify_storage_s3_dart/src/storage_s3_service/storage_s3_service.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+
+/// The io implementation of `downloadFile` API.
+@internal
+S3DownloadFileOperation downloadFile({
+  required StorageDownloadFileRequest request,
+  required S3PluginConfig s3pluginConfig,
+  required StorageS3Service storageS3Service,
+  required AppPathProvider appPathProvider,
+  void Function(S3TransferProgress)? onProgress,
+}) {
+  final s3Options = request.options as S3DownloadFileOptions? ??
+      S3DownloadFileOptions(
+        accessLevel: s3pluginConfig.defaultAccessLevel,
+      );
+  final targetIdentityId = s3Options.targetIdentityId;
+  final downloadDataOptions = targetIdentityId == null
+      ? S3DownloadDataOptions(
+          accessLevel: s3Options.accessLevel,
+          getProperties: s3Options.getProperties,
+        )
+      : S3DownloadDataOptions.forIdentity(
+          targetIdentityId,
+          getProperties: s3Options.getProperties,
+        );
+
+  late final String destinationPath;
+  late final IOSink sink;
+  late final File tempFile;
+
+  final downloadDataTask = storageS3Service.downloadData(
+    key: request.key,
+    options: downloadDataOptions,
+    // Ensure destination file is writable. Exception thrown in the check
+    // will be forwarded to the Future, downloadDataTask.result below
+    preStart: () async {
+      destinationPath = await _ensureDestinationWritable(request.localFile);
+      tempFile = File(
+        path.join(
+          await appPathProvider.getTemporaryPath(),
+          'amplify_storage_s3_temp_${uuid()}',
+        ),
+      );
+      sink = tempFile.openWrite(mode: FileMode.append);
+    },
+    onProgress: onProgress,
+    // ignore: unnecessary_lambdas
+    onData: (bytes) {
+      sink.add(bytes);
+    },
```

Comment:
```suggestion
    onData: sink.add,
```

Reply:
Explained in the original PR, sink is a late reference which gets set in `preStart` callback. `preStart` is guaranteed to run and set the sink by the implementation.

When use `onData: sink.add,` as sink is late, so it would throw

```
preventLateInitializationError: Local 'sink' has not been initialized.
```

I will add a comment here to explain the reason (meant to do so but forgot :P)

Reply:
Ah, yeah I remember now. Thanks for adding a comment.

Reply:
In general, any `ignore` comment should have an explanation. I'll make sure I do the same as well.